### PR TITLE
Port BT Navigator to lifecycle nodes

### PR DESCRIPTION
**Pull request by mjeronimo**
## Description

* Port the BT Navigator to lifecycle nodes
* Implement a ROS2 Action server instead of task server

## Future work

* The task clients still need to be converted to action clients once the other task servers are converted to action servers

### CODE REVIEW

File: nav2_bt_navigator/src/bt_navigator.cpp [hunk 1]
```diff
@@ -12,76 +12,207 @@
...
+      // to their corresponding task servers
+      bt_->cancelAllActions(tree_->root_node);
+
+      // Reset the BT so that it can be run again in the future
+      bt_->resetTree(tree_->root_node);
+      goal_handle->set_canceled(result);
+      break;
+
+    case TaskStatus::SUCCEEDED:
+      goal_handle->set_succeeded(result);
```
Comment (bpwilcox) [inline nav2_bt_navigator/src/bt_navigator.cpp]: Action state transitions have been renamed in master branch of rclcpp_action, `goal_handle->set_succeeded` is now `goal_handle->succeed`

File: nav2_bt_navigator/src/bt_navigator.cpp [hunk 2]
```diff
@@ -12,76 +12,207 @@
...
+  switch (rc) {
+    case TaskStatus::CANCELED:
+      // Even though the BT is no longer running, remote actions may still be executing. So,
+      // an explicit canceling of all actions allows the task clients to send cancel messages
+      // to their corresponding task servers
+      bt_->cancelAllActions(tree_->root_node);
+
+      // Reset the BT so that it can be run again in the future
+      bt_->resetTree(tree_->root_node);
+      goal_handle->set_canceled(result);
```
Comment (bpwilcox) [inline nav2_bt_navigator/src/bt_navigator.cpp]: Action state transitions have been renamed in master branch of rclcpp_action, `goal_handle->set_canceled` is now `goal_handle->canceled`

File: nav2_bt_navigator/src/bt_navigator.cpp [hunk 3]
```diff
@@ -12,76 +12,207 @@
...
+      break;
+
+    case TaskStatus::SUCCEEDED:
+      goal_handle->set_succeeded(result);
+      break;
+
+    case TaskStatus::FAILED:
+      goal_handle->set_aborted(result);
```
Comment (bpwilcox) [inline nav2_bt_navigator/src/bt_navigator.cpp]: Action state transitions have been renamed in master branch of rclcpp_action, `goal_handle->set_aborted` is now `goal_handle->abort`

File: nav2_bt_navigator/src/bt_navigator.cpp [hunk 4]
```diff
@@ -12,76 +12,207 @@
...
+  tree_->nodes = std::move(temp_tree.nodes);
+  temp_tree.root_node = nullptr;
 
-  // Create and run the behavior tree
-  NavigateToPoseBehaviorTree bt(shared_from_this());
+  return nav2_lifecycle::CallbackReturn::SUCCESS;
+}
+
+nav2_lifecycle::CallbackReturn
+BtNavigator::on_activate(const rclcpp_lifecycle::State &)
```
Comment (mhpanah) [inline nav2_bt_navigator/src/bt_navigator.cpp]: should we add `/*state*/` to be consistent?  similar to `BtNavigator::on_configure(const rclcpp_lifecycle::State & /*state*/)`

File: nav2_bt_navigator/src/bt_navigator.cpp [hunk 5]
```diff
@@ -12,76 +12,207 @@
...
+{
+  RCLCPP_INFO(get_logger(), "Activating");
+
+  return nav2_lifecycle::CallbackReturn::SUCCESS;
+}
+
+nav2_lifecycle::CallbackReturn
+BtNavigator::on_deactivate(const rclcpp_lifecycle::State &)
```
Comment (mhpanah) [inline nav2_bt_navigator/src/bt_navigator.cpp]: Same as above.

File: nav2_bt_navigator/src/bt_navigator.cpp [hunk 6]
```diff
@@ -12,76 +12,207 @@
...
+{
+  RCLCPP_INFO(get_logger(), "Deactivating");
+
+  return nav2_lifecycle::CallbackReturn::SUCCESS;
+}
 
-  TaskStatus result = bt.run(blackboard, xml_string,
-      std::bind(&nav2_tasks::NavigateToPoseTaskServer::cancelRequested, task_server_.get()));
+nav2_lifecycle::CallbackReturn
+BtNavigator::on_cleanup(const rclcpp_lifecycle::State &)
```
Comment (mhpanah) [inline nav2_bt_navigator/src/bt_navigator.cpp]: Same as above.

File: nav2_bt_navigator/src/bt_navigator.cpp [hunk 7]
```diff
@@ -12,76 +12,207 @@
...
+  xml_string_.clear();
+  tree_.reset();
+  blackboard_.reset();
+  bt_.reset();
+
+  return nav2_lifecycle::CallbackReturn::SUCCESS;
+}
+
+nav2_lifecycle::CallbackReturn
+BtNavigator::on_error(const rclcpp_lifecycle::State &)
```
Comment (mhpanah) [inline nav2_bt_navigator/src/bt_navigator.cpp]: same as above.

File: nav2_bt_navigator/src/bt_navigator.cpp [hunk 8]
```diff
@@ -12,76 +12,207 @@
...
+{
+  RCLCPP_ERROR(get_logger(), "Handling error state");
+  return nav2_lifecycle::CallbackReturn::SUCCESS;
+}
+
+nav2_lifecycle::CallbackReturn
+BtNavigator::on_shutdown(const rclcpp_lifecycle::State &)
```
Comment (mhpanah) [inline nav2_bt_navigator/src/bt_navigator.cpp]: same as above.

File: nav2_bt_navigator/src/navigate_to_pose_behavior_tree.cpp [hunk 9]
```diff
@@ -71,25 +76,24 @@ BT::NodeStatus NavigateToPoseBehaviorTree::updatePath(BT::TreeNode & tree_node)
...
+BT::NodeStatus
+NavigateToPoseBehaviorTree::initialPoseReceived(BT::TreeNode & tree_node)
 {
   auto initPoseReceived = tree_node.blackboard()->template get<bool>("initial_pose_received");
-  if (initPoseReceived) {
-    return BT::NodeStatus::SUCCESS;
-  }
-  return BT::NodeStatus::FAILURE;
+  return initPoseReceived ? BT::NodeStatus::SUCCESS : BT::NodeStatus::FAILURE;
 }
```
Comment (mhpanah) [inline nav2_bt_navigator/src/navigate_to_pose_behavior_tree.cpp]: In master there is a new simple action node for clearing the entire costmap.

File: nav2_bt_navigator/src/bt_navigator.cpp (hunk 1 above)
Comment (mjeronimo) [reply nav2_bt_navigator/src/bt_navigator.cpp]: Fixed.

File: nav2_bt_navigator/src/bt_navigator.cpp (hunk 2 above)
Comment (mjeronimo) [reply nav2_bt_navigator/src/bt_navigator.cpp]: Fixed.

File: nav2_bt_navigator/src/bt_navigator.cpp (hunk 3 above)
Comment (mjeronimo) [reply nav2_bt_navigator/src/bt_navigator.cpp]: Fixed.

File: nav2_bt_navigator/src/bt_navigator.cpp (hunk 4 above)
Comment (mjeronimo) [reply nav2_bt_navigator/src/bt_navigator.cpp]: Fixed.

File: nav2_bt_navigator/src/bt_navigator.cpp (hunk 6 above)
Comment (mjeronimo) [reply nav2_bt_navigator/src/bt_navigator.cpp]: Fixed.

File: nav2_bt_navigator/src/bt_navigator.cpp (hunk 8 above)
Comment (mjeronimo) [reply nav2_bt_navigator/src/bt_navigator.cpp]: Fixed.

File: nav2_bt_navigator/src/navigate_to_pose_behavior_tree.cpp (hunk 9 above)
Comment (mjeronimo) [reply nav2_bt_navigator/src/navigate_to_pose_behavior_tree.cpp]: Ok, thanks. We can include that node, if desired, in BTs.

**Pull request merged**